ssl/quic/quic_impl.c

```diff
@@ -9,271 +9,1241 @@
+{
+    /* Cannot be changed after handshake started */
+    if (qc->started)
+        return;
+
+    qc->as_server = 0;
+}
+
+/* SSL_set_accept_state */
+void ossl_quic_set_accept_state(QUIC_CONNECTION *qc)
+{
+    /* Cannot be changed after handshake started */
+    if (qc->started)
+        return;
+
+    qc->as_server = 1;
+}
+
+/* SSL_do_handshake */
+struct quic_handshake_wait_args {
+    QUIC_CONNECTION     *qc;
+};
+
+static int quic_handshake_wait(void *arg)
+{
+    struct quic_handshake_wait_args *args = arg;
+
+    if (!ossl_quic_channel_is_active(args->qc->ch))
+        return -1;
+
+    if (ossl_quic_channel_is_handshake_complete(args->qc->ch))
+        return 1;
+
     return 0;
 }
 
-long ossl_quic_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
+static int configure_channel(QUIC_CONNECTION *qc)
 {
-    switch(cmd) {
-    case SSL_CTRL_CHAIN:
-        if (larg)
-            return ssl_cert_set1_chain(NULL, ctx, (STACK_OF(X509) *)parg);
-        else
-            return ssl_cert_set0_chain(NULL, ctx, (STACK_OF(X509) *)parg);
+    assert(qc->ch != NULL);
 
-    case SSL_CTRL_SET_TLSEXT_TICKET_KEYS:
-    case SSL_CTRL_GET_TLSEXT_TICKET_KEYS:
-        /* TODO(QUIC): these will have to be implemented properly */
+    if (!ossl_quic_channel_set0_net_rbio(qc->ch, qc->net_rbio)
+        || !ossl_quic_channel_set0_net_wbio(qc->ch, qc->net_wbio)
+        || !ossl_quic_channel_set_peer_addr(qc->ch, &qc->init_peer_addr))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * Creates a channel and configures it with the information we have accumulated
+ * via calls made to us from the application prior to starting a handshake
+ * attempt.
+ */
+static int ensure_channel_and_start(QUIC_CONNECTION *qc)
+{
+    QUIC_CHANNEL_ARGS args = {0};
+
+    if (qc->ch != NULL)
         return 1;
+
+    args.libctx     = qc->ssl.ctx->libctx;
+    args.propq      = qc->ssl.ctx->propq;
+    args.is_server  = 0;
+
+    qc->ch = ossl_quic_channel_new(&args);
+    if (qc->ch == NULL)
+        return 0;
+
+    if (!configure_channel(qc)
+        || !ossl_quic_channel_start(qc->ch)) {
+        ossl_quic_channel_free(qc->ch);
+        qc->ch = NULL;
+        return 0;
     }
-    return 0;
+
+    qc->stream0 = ossl_quic_channel_get_stream_by_id(qc->ch, 0);
+    if (qc->stream0 == NULL) {
+        ossl_quic_channel_free(qc->ch);
+        qc->ch = NULL;
+        return 0;
+    }
+
+    qc->started = 1;
+    return 1;
 }
 
-long ossl_quic_callback_ctrl(SSL *s, int cmd, void (*fp) (void))
+int ossl_quic_do_handshake(QUIC_CONNECTION *qc)
 {
-    return 0;
+    int ret;
+
+    fprintf(stderr, "# handshake begin\n");
+
+    if (qc->ch != NULL && ossl_quic_channel_is_term_any(qc->ch))
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, SSL_R_PROTOCOL_IS_SHUTDOWN, NULL);
+
+    if (BIO_ADDR_family(&qc->init_peer_addr) == AF_UNSPEC)
+        /* Peer address must have been set. */
+        return QUIC_RAISE_NON_NORMAL_ERROR(qc, ERR_R_PASSED_INVALID_ARGUMENT, NULL);
```

Comment:
When trying to fuzz it, I'm not getting past this. I'm not sure how to best avoid this. The fuzzer just receives data from a BIO, and sends data to a BIO, and that BIO doesn't have an address. I'm not sure we want to support not knowing what the peer address is is normal operations.

Reply:
The BIO needs to have datagram semantics. You can try to use BIO_dgram_pair, though while unit tested I haven't verified its interactions with this PR yet. BIO_dgram_pair is designed to support passing network addresses as metadata of each datagram, so you can just make up a non-zero dummy address to use with it. See the manpage for more details.

Reply:
I've tried changing it to BIO_s_dgram_mem(), which seems like the BIO_s_mem() replacement. However, I currently don't see a way to have that set an address.

Reply:
`BIO_s_dgram_mem()` doesn't support that yet (I didn't need it for the purpose that I wrote it for)...we should add that - as well as support for setting MTU values.

Reply:
BIO_s_dgram_mem provides a unidirectional stream... BIO_s_dgram_pair provides a bidirectional one. It's possible BIO_s_dgram_mem doesn't implement this functionality since I believe it was created to fulfill an internal need. So I'd recommend using BIO_s_dgram_pair here.

Reply:
I don't need a pair. I have an file from which I read. I currently get a size and the data from that. It could be changed to also read an address if needed. A BIO_s_dgram_mem() looks good enough for that. For writing, I don't care where things go to, nothing is going to be done with that data.